src/integration-test/java/com/commercetools/sync/integration/services/impl/CartDiscountServiceImplIT.java

```diff
@@ -0,0 +1,432 @@
+        final SphereClient spyClient = spy(CTP_TARGET_CLIENT);
+        when(spyClient.execute(any(CartDiscountQuery.class)))
+                .thenReturn(exceptionallyCompletedFuture(new BadGatewayException()))
+                .thenCallRealMethod();
+
+        final CartDiscountSyncOptions spyOptions =
+                CartDiscountSyncOptionsBuilder.of(spyClient)
+                        .errorCallback((errorMessage, exception) -> {
+                            errorCallBackMessages.add(errorMessage);
+                            errorCallBackExceptions.add(exception);
+                        })
+                        .build();
+
+        final CartDiscountService spyCartDiscountService = new CartDiscountServiceImpl(spyOptions);
+
+
+        final Set<String> cartDiscountKeys = new HashSet<>();
+        cartDiscountKeys.add(CART_DISCOUNT_KEY_1);
+
+        // test and assert
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(spyCartDiscountService.fetchMatchingCartDiscountsByKeys(cartDiscountKeys))
+                .hasFailedWithThrowableThat()
+                .isExactlyInstanceOf(BadGatewayException.class);
+    }
+
+    @Test
+    void fetchCachedCartDiscountId_WithNotCachedExistingCartDiscount_ShouldFetchCartDiscount() {
+        final Optional<String> cartDiscountId = cartDiscountService.fetchCachedCartDiscountId(CART_DISCOUNT_KEY_1)
+                .toCompletableFuture()
+                .join();
+
+        assertThat(cartDiscountId).isNotEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchCachedCartDiscountId_WithBlankCartDiscountKey_ShouldReturnEmptyOptional() {
+        final Optional<String> cartDiscountId = cartDiscountService.fetchCachedCartDiscountId("  ")
+                .toCompletableFuture()
+                .join();
+
+        assertThat(cartDiscountId).isEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchMatchingCartDiscountsByKeys_WithAllExistingSetOfKeys_ShouldCacheFetchedCartDiscountIds() {
+        //preparation
+        final Set<CartDiscount> fetchedCartDiscounts =
+                cartDiscountService.fetchMatchingCartDiscountsByKeys(singleton(CART_DISCOUNT_KEY_1))
+                        .toCompletableFuture().join();
+        assertThat(fetchedCartDiscounts).hasSize(1);
+
+        final Optional<CartDiscount> cartDiscountOptional = CTP_TARGET_CLIENT
+                .execute(CartDiscountQuery.of()
+                        .withPredicates(queryModel -> queryModel.name().lang(Locale.ENGLISH).is(CART_DISCOUNT_KEY_1)))
+                .toCompletableFuture()
+                .join()
+                .head();
+
+        assertThat(cartDiscountOptional).isNotNull();
```

Comment:
why do we need this?

Reply:
5cdd3f9de578f817bda650bce4b2d122c41f0c03